d: Add an easy way to use D-specific features

ximion opened the pull request:
Meson had a D-specific `unittest_args()` method for a while to have a compiler-agnostic way of toggling embedded unit-tests on and off.

This patch removes that method and replaces it with a more generic one (that may also be used by other compilers) supporting more of D's features.
The features supported are:
 * unittest: Allows switching embedded unit-tests on and off, supporting the default way D programmers write tests
 * versions: Allows specifying feature versions (similar to C's preprocessor macros for conditional compilation) to create different versions of a D program, which is D's default and only way for doing conditional compilation.
 * import_dirs: Used to set directories for string imports. Those are usually used for D's advanced templates, and are especially used for in the Vibe.d web-framework for templates.

Supporting all of these in Meson does make sense, because they are incredibly common when using D (I deliberately left out the more obscure features like GC-allocation-tracing for now).

Of course D compilers have different flags to set all of these options, having Meson abstract that in one place therefore makes a lot of sense instead of copying boilerplate code around.

This patch removes the previous `unittest_args()` method, which will break projects using it. Reason for its removal is that it's rendered pointless by the new `feature_args()` method and would only be dead weight. It is also not documented properly, and it is very easy for existing projects to switch to the new version.
If compatibility is wanted anyway, I would extend this patch with some small compatibility code (I am torn on whether removing the old method is a good idea - if we keep it, we will probably have it for eternity).

If this patch is accepted, I intent to document this feature in some D specific section of the Meson documentation.

Thanks!
    Matthias

ignatenkobrain commented on mesonbuild/compilers/d.py:
```diff
@@ -27,6 +27,21 @@
 )
 
+d_feature_args = {
+                  'gcc':  {'unittest': '-funittest',
```
__[flake8]__
* [E126] continuation line over-indented for hanging indent

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/compilers/d.py:
```diff
@@ -27,6 +27,21 @@
+                           'version': '-version',
+                           'import_dir': '-J'
+                           }
+                 }
```
__[flake8]__
* [E126] continuation line over-indented for hanging indent

<!-- Comment by [SideCI](https://sideci.com) -->

TingPing commented on test cases/d/9 features/meson.build:
```diff
@@ -0,0 +1,28 @@
+dc = meson.get_compiler('d')
+
+# directory for data
+data_dir = join_paths(meson.current_source_dir(), 'data')
```
Passing `'data'` should work.

ximion replied:
In that case I will have to expand the string depending on the context. I will look into that.

ximion replied:
Okay, the path handling works like the rest of Meson now (including both the path in the build directory and the source directory in case a relative path is given).

ignatenkobrain commented on mesonbuild/compilers/d.py:
```diff
@@ -79,8 +93,50 @@ def get_soname_args(self, prefix, shlib_name, suffix, path, soversion, is_shared
+                    res.append('{0}{1}'.format(import_dir_arg, os.path.join(env.build_dir, d)))
+
+
+        if kwargs:
```
__[flake8]__
* [E303] too many blank lines (2)

<!-- Comment by [SideCI](https://sideci.com) -->

jpakkane commented:
Can you use internal unittests also in library targets? If yes, how are they run?

ximion commented:
@jpakkane You can. The compiler will collect all the `unittest { }` blocks and glue them together, then create a dummy `main()` function to execute the tests.

Some compilers (LDC/DMD) support a `-main` flag to explicitly call for the generation of a dummy main function, even for executables (overriding the real main), while GDC does not support this. Many developers also use a `version (Unittest)` directive to compile a specific main function for unittests. This also works for libraries (with no main being added normally, and a dedicated main function being included when the tests are compiled).

jpakkane commented:
But suppose you build a `libfoo.so`. How would you invoke the tests on it? With a special executable? Does it also generate a `libfoo_test` executable next to the library? Something else?

ximion commented:
Ah, yes - you need a special executable. You basically compile your sources twice, once without the unittest flag, which yields the shared library, and once with it, which yields an executable.

I have never attempted to supply both the flags needed for building a shared library and the unittest flag... I guess an attempt like that will fail with a compile error.

jpakkane commented:
So if you do this:

    lib = shared_library(...)
    exe = executable(..., link_with : lib)

and enable unit test mode, it will fail because the output of the compilation command is not a library but instead an executable?

ximion commented:
No, it won't fail, because the unittest-mode is no global flag.
Consider:
```
lib_src = ['aaa.d', 'bbb.d']
app_src = ['main.d']

lib = shared_library(lib_src)
exe = executable(app_src, link_with : lib)
```
This will build an executable and a shared library, no tests involved.

```
dc = meson.get_compiler('d')
unittest_arg = dc.feature_args(unittest: true)
test_exe = executable(lib_src, d_args: unittest_arg)
test('foo', test_exe)
```
This will build a test executable for the library only.

```
lib_src = ['aaa.d', 'bbb.d']
app_src = ['main.d']
dc = meson.get_compiler('d')
unittest_arg = dc.feature_args(unittest: true)

lib = shared_library(lib_src)
exe = executable(app_src, link_with: lib)

test_exe = executable(lib_src + app_src, d_args: unittest_arg)
test('foo', test_exe)
```
This will build the library, the executable, and a test executable for both the lib and exe (that contains all the unittest scopes in the files aaa.d, bbb.d and main.d)

jpakkane commented:
But that means compiling all source code twice always. This is exactly the same thing what Rust does. They handwave it away by having `cargo test` build only test code and run it but the principle is the same. In D the issue is not so grave because it compiles so fast but not so for Rust. Still, it makes writing build system description for these things complicated because they have optimized for their own use case and not cooperation with others.

ximion commented:
> But that means compiling all source code twice always

Yes, and that is annoying sometimes - in some occasions, I would prefer that compilation to only run when I run `ninja test`, but that's something to think about for later.

> In D the issue is not so grave because it compiles so fast but not so for Rust

D compilation speed also gets down to slower levels as soon as you use an optimizing compiler (GDC/LDC) and enable stronger optimizations. It's not as crazy as with Rust though.

> Still, it makes writing build system description for these things complicated because they have optimized for their own use case and not cooperation with others.

If this is a common issue, we can maybe do something in order to support it better in Meson (e.g. by compiling only when `ninja test` is called). Some languages are pretty hardwired to their own tooling, I haven't used Rust without cargo yet, for example, or Go without its own buildsystem. For D, while it also works very well with its own buildsystem dub, the coupling is fortunately not incredibly tight.

Personally, I use the unittest blocks in D for smaller functions or when I write functional-style code, because in that case they are incredibly convenient, and create a dedicated separate testsuite for the remaining cases (especially because the software I write requires larger amounts of data samples, and integrating that with unittest blocks is not that nice).

ximion commented:
It turned out the unittest_args)` property was used much more than I expected, therefore I added some small compatibility function, so we can remove the method later after most projects have switched to the more generic variant.

ximion commented:
Any feedback on this patch? Is this okay to merge, or are there any changes needed? I know it's not the most ideal thing, but since `get_unittest_args` was in there before, which was even less ideal, replacing it would make much sense to me.

jpakkane commented:
Sorry, I've been busy with other features and the bugfix release. I'll do a full review of outstanding MRs this weekend.

jpakkane commented:
We'll probably want to have something like this but for all languages. For example to specify new C++ features such as using the file system TS (which requires extra compiler arguments currently). This means adding both compiler and linker flags. Maybe we could have a better experience for this by doing something like:

    executable(..., features : ['d_unittest'])

or 

    executable(..., features : ['cpp_filesystem'])

This way people would not need to extract compiler flags and things from the object and pass them around, but instead could directly declare what they want.

The syntax presented is a bit crap, but would something like this make sense?

ximion commented:
> The syntax presented is a bit crap, but would something like this make sense?

Yes, I think that would be even better :-) I also think the syntax is actually fine... I can try to implement something like this.

ximion commented:
On second thought: Some of the arguments used for compiling should also end up in the generated pkg-config files. With the existing solution, that was easily possible, by adding a `features` property to targets, that will become harder. So I think having both functions makes sense.

ximion commented:
@jpakkane I tried to implement this, but failed because Meson doesn't know a dictionary type in build definitions. So something like:
```
executable(..., d_features : ['versions': ['Derelict_Static']])
```
fails. I could of course fake it by having something like `['version::Derelict_Static', 'version::Have_64bit_timestamp', ...]` but that feels really ugly.

nirbheek commented:
Can you do `d_feature_versions : ['Derelict_Static', 'Have_64bit_timestamp']` et al or is the list of possible features too large to be enumerated?

ximion commented:
@nirbheek Versions are defined by the project, and there is an arbitrary number of them (but I would argue insanity starts if you have more then 20).
The problem with `d_feature_versions : ['Derelict_Static', 'Have_64bit_timestamp']` is that you would need a `d_feature_${feature}` property for each and every feature that exists. At the moment, that would just be `d_feature_unittest: bool`, `d_feature_versions: list` and `d_feature_import_dirs: list`, but who knows what will be added in future and - if/when other languages use this as well - how many of these properties we will have.
That being said, the number of properties will never be crazy high (the three ones are all I can think of for D at time), and other languages likewise would not have insane numbers of features. Also, it is a rather nice way to represent these things.
Ultimately, it's a maintainer's design decision here, I will implement whatever is seen as the best fit :-)

nirbheek commented:
I think it's fine to implement just those three kwargs for now. If the number of features explodes, we will have a good reason to implement a dictionary/associative array type in the parser and add a generic `d_features:` kwarg.

ximion commented:
I implemented the kwargs and kept the `get_feature_args` method as well, so I can pass the arguments to other tools (pkg-config in particular, where you really want version flags to be present sometimes).

I have moved the `process_compilers()` call in target initialization up before `process_kwargs()` is called, because I need access to the compiler to fetch specific flags. This should not cause any problems, looking at the code, and it appears that all tests are running fine as well.

ximion commented:
I changed my mind on `get_feature_args` - the only thing I would really want it for is passing it to `pkgconfig.generate`, and in that case I only want the versions feature. The unittest feature is completely useless in pkgconfig, and while I can fabricate cases where using the `import_dir` with a pkg-config file makes some sense, I would think it's generally not a great idea to use it that way and I know no project which actually does that.
Therefore, I just added a `d_feature_versions` kwark to `pkgconfig.generate` and dropped `get_feature_args` completely (if you want I can of course squash all commits into one before merge).

ximion commented:
Any comment on the current implementation? (sorry for nagging, but I have a project where I would like to replace the compiler-specific hardcoding and submit the changes back upstream)

jpakkane commented:
I can't think of a simple way to make this general to all languages. Let's add this in now, we can transition to something better later if we can think of one. Please update the documentation to match.

ximion commented:
@jpakkane 
> I can't think of a simple way to make this general to all languages

In that case I've given the properties slightly less verbose names (they are prefixed with d_ already anyway).
The Meson D support is fully documented now as well :-)

ximion commented:
Thank you very much :-)
